compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignConnectionTypes.cpp

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+
+#define DEBUG_TYPE "iree-amdaie-assign-connection-types"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+class AMDAIEAssignConnectionTypesPass
+    : public impl::AMDAIEAssignConnectionTypesBase<
+          AMDAIEAssignConnectionTypesPass> {
+ public:
+  AMDAIEAssignConnectionTypesPass() = default;
+  AMDAIEAssignConnectionTypesPass(
+      const AMDAIEAssignConnectionTypesPass &pass){};
+  AMDAIEAssignConnectionTypesPass(
+      const AMDAIEAssignConnectionTypesOptions &options)
+      : AMDAIEAssignConnectionTypesBase(options) {}
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+void AMDAIEAssignConnectionTypesPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  IRRewriter rewriter(parentOp->getContext());
+  // Assign connection types based on provided options.
+  WalkResult res = parentOp->walk([&](AMDAIE::ConnectionOp connectionOp) {
+    AMDAIE::ConnectionType connectionType =
+        enablePacketFlow ? AMDAIE::ConnectionType::Packet
+                         : AMDAIE::ConnectionType::Circuit;
+    auto sourceLogicalObjFifo =
+        dyn_cast_if_present<AMDAIE::LogicalObjFifoOpInterface>(
+            connectionOp.getSource().getDefiningOp());
+    auto targetLogicalObjFifo =
+        dyn_cast_if_present<AMDAIE::LogicalObjFifoOpInterface>(
+            connectionOp.getTarget().getDefiningOp());
+    // TODO(jornt): only enable for connections from/to L3 for now, generalize
+    // later as there are some issues for some matmul shapes.
+    if ((sourceLogicalObjFifo &&
+         sourceLogicalObjFifo.getMemorySpaceAsUInt() == 0) ||
+        (targetLogicalObjFifo &&
+         targetLogicalObjFifo.getMemorySpaceAsUInt() == 0)) {
+      rewriter.setInsertionPoint(connectionOp);
+      rewriter.replaceOpWithNewOp<AMDAIE::ConnectionOp>(
+          connectionOp, connectionOp.getTarget(),
+          connectionOp.getTargetChannels(), connectionOp.getSource(),
+          connectionOp.getSourceChannels(),
+          ConnectionTypeAttr::get(rewriter.getContext(), connectionType));
```

Comment:
Likewise. You can have this defined before the walk and simply reuse.

Reply:
I added an else so kept this inside the walk. Could you check again?

Reply:
```
    if ((sourceLogicalObjFifo &&
         sourceLogicalObjFifo.getMemorySpaceAsUInt() == 0) ||
        (targetLogicalObjFifo &&
         targetLogicalObjFifo.getMemorySpaceAsUInt() == 0)) {
      connectionAttr =
          ConnectionTypeAttr::get(rewriter.getContext(), connectionType);
    } else {
      connectionAttr = ConnectionTypeAttr::get(rewriter.getContext(),
                                               AMDAIE::ConnectionType::Circuit);
    }
```
I'd still advocate on pulling both of these out of the walk - form the attribute once and reuse it later depending on the condition.

So something like, `connectionAttr = canApplyPacketFlow(...) ? packetAttr : circuitAttr`.
And in that essence you may create a named lambda function (or a static utility) `canApplyPacketFlow` (I'm not good with names).

Reply:
Sure, I pulled it out